### PR TITLE
vDPA High Availability

### PR DESCRIPTION
**Pull request by Ch3n60x**
This PR introduces vDPA high availability.

### CODE REVIEW

File: app/virtio-ha/main.c [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+	return HA_MSG_HDLR_REPLY;
+}
+
+static int
+ha_server_pf_store_ctx(struct virtio_ha_msg *msg)
+{
+	struct virtio_ha_pf_dev_list *list = &hs.pf_list;
+	struct virtio_ha_pf_dev *dev;
+
+	if (msg->nr_fds != 2)
```
Comment (kailiangz1) [inline app/virtio-ha/main.c:193]: should return error in this situation

File: app/virtio-ha/main.c [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+
+	TAILQ_FOREACH(dev, list, next) {
+		if (!strcmp(dev->pf_name.dev_bdf, msg->hdr.bdf)) {
+			vf_list = &dev->vf_list;
+			found = true;
+			break;
+		}
+	}
+
+	if (!found)
```
Comment (kailiangz1) [inline app/virtio-ha/main.c:280]: should return err here also

File: drivers/common/virtio_mi/lm.c [hunk 3]
```diff
@@ -1155,6 +1163,26 @@ virtio_vdpa_mi_dev_remove(struct rte_pci_device *pci_dev)
 	return 0;
 }
 
+static void
+virtio_ha_pf_drv_ctx_set(const struct virtio_dev_name *pf, const void *ctx)
+{
+	const struct virtio_pf_ctx *pf_ctx = (const struct virtio_pf_ctx *)ctx;
+
+	memcpy(&cached_ctx.pf_name, pf, sizeof(struct virtio_dev_name));
+	cached_ctx.vfio_group_fd = pf_ctx->vfio_group_fd;
```
Comment (kailiangz1) [inline drivers/common/virtio_mi/lm.c:1232]: how about use dev args to pass parameter? 
like in rte_eal_hotplug_add("pci", pf_name, "vdpa=2");

File: app/vfe-vdpa/main.c [hunk 4]
```diff
@@ -727,6 +728,113 @@ cmdline_parse_ctx_t main_ctx[] = {
...
+			ret = virtio_ha_vf_ctx_query(&vf_list[j].vf_name, pf_list + i, &vf_ctx);
+			if (ret < 0) {
+				RTE_LOG(ERR, VDPA, "Failed to query vf ctx\n");
+				return -1;
+			}
+
+			ret = virtio_ha_vf_ctx_set(&vf_list[j].vf_name, vf_ctx);
+			if (ret < 0) {
+				RTE_LOG(ERR, VDPA, "Failed to set vf ctx in vf driver\n");
+				return -1;
```
Comment (kailiangz1) [inline app/vfe-vdpa/main.c]: should free vf_ctx and vf_list if failed

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 5]
```diff
@@ -835,50 +852,57 @@ virtio_vdpa_dev_set_mem_table(int vid)
 		return ret;
 	}
 
+	for (i = 0; i < cur_mem->nregions; i++) {
+		vhost_reg = &cur_mem->regions[i];
+		host_phys_addrs[i] = rte_mem_virt2phy((void *)(uintptr_t)vhost_reg->host_user_addr);
+		if (host_phys_addrs[i] == RTE_BAD_IOVA) {
+			DRV_LOG(ERR, "virt2phy translate failed");
+			return -1;
```
Comment (kailiangz1) [inline drivers/vdpa/virtio/virtio_vdpa.c:938]: free(cur_mem);

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 6]
```diff
@@ -1336,6 +1473,46 @@ virtio_vdpa_dev_config(int vid)
...
+			return 0;
+		}
+
+		ret = virtio_ha_vf_vhost_fd_store(&priv->vf_name, &priv->pf_name, vhost_sock_fd);
+		if (ret) {
+			DRV_LOG(ERR, "Failed to store vhost fd (vid %d)", vid);
+			return 0;			
+		}
+
+		priv->ctx_stored = true;
```
Comment (kailiangz1) [inline drivers/vdpa/virtio/virtio_vdpa.c:1518]: priv->ctx_stored should be flase after dev_close

File: drivers/common/virtio_mi/lm.c (hunk 3 above)
Comment (Ch3n60x) [reply drivers/common/virtio_mi/lm.c:1232]: yes it's an option and was discussed before. Parav does not want to add more devargs. I am fine with either option.

File: app/virtio-ha/main.c (hunk 1 above)
Comment (Ch3n60x) [reply app/virtio-ha/main.c:193]: done

File: app/virtio-ha/main.c (hunk 2 above)
Comment (Ch3n60x) [reply app/virtio-ha/main.c:280]: done

File: app/vfe-vdpa/main.c (hunk 4 above)
Comment (Ch3n60x) [reply app/vfe-vdpa/main.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 5 above)
Comment (Ch3n60x) [reply drivers/vdpa/virtio/virtio_vdpa.c:938]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 6 above)
Comment (Ch3n60x) [reply drivers/vdpa/virtio/virtio_vdpa.c:1518]: done

**Pull request merged**